pkg/agent/controller/trafficcontrol/controller.go

```diff
@@ -0,0 +1,1037 @@
+		var namespaces []*v1.Namespace
+		nsSelector, err = metav1.LabelSelectorAsSelector(appliedTo.NamespaceSelector)
+		if err != nil {
+			return nil, err
+		}
+		namespaces, err = c.namespaceLister.List(nsSelector)
+		if err != nil {
+			return nil, err
+		}
+		// Select Pods with Pod selector from all selected namespaces.
+		for _, ns := range namespaces {
+			pods, err := c.podLister.Pods(ns.Name).List(podSelector)
+			if err != nil {
+				return nil, err
+			}
+			selectedPods = append(selectedPods, pods...)
+		}
+	} else {
+		// If Namespace selector is nil, use Pod selector to select Pods from all namespaces.
+		selectedPods, err = c.podLister.List(podSelector)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var nonHostNetworkPods []*v1.Pod
+	// TrafficControl does not support host network Pods.
+	for _, pod := range selectedPods {
+		if !pod.Spec.HostNetwork {
+			nonHostNetworkPods = append(nonHostNetworkPods, pod)
+		}
+	}
+
+	return nonHostNetworkPods, nil
+}
+
+func genTunnelPortName(portNamePrefix string, config interface{}) string {
+	hash := sha1.New() // #nosec G401: not used for security purposes
+	buf := make([]byte, 4)
+
+	writeUint32 := func(val uint32) {
+		binary.BigEndian.PutUint32(buf, val)
+		hash.Write(buf)
+	}
+	writeIP := func(ip string) {
+		hash.Write(net.ParseIP(ip))
+	}
+
+	switch config.(type) {
+	case *v1alpha2.UDPTunnel:
+		tunnelConfig := config.(*v1alpha2.UDPTunnel)
+		var dstPort, vni uint32
+		if tunnelConfig.DestinationPort != nil {
+			dstPort = uint32(*tunnelConfig.DestinationPort)
+		}
+		if tunnelConfig.VNI != nil {
+			vni = uint32(*tunnelConfig.VNI)
+		}
+		writeIP(tunnelConfig.RemoteIP)
+		writeUint32(dstPort)
+		writeUint32(vni)
+
+	case *v1alpha2.GRETunnel:
+		tunnelConfig := config.(*v1alpha2.GRETunnel)
+		writeIP(tunnelConfig.RemoteIP)
+		if tunnelConfig.Key != nil {
+			writeUint32(uint32(*tunnelConfig.Key))
+		}
+
+	case *v1alpha2.ERSPANTunnel:
+		tunnelConfig := config.(*v1alpha2.ERSPANTunnel)
+		var sessionID, index, dir, hardwareID uint32
+		if tunnelConfig.SessionID != nil {
+			sessionID = uint32(*tunnelConfig.SessionID)
+		}
+		if tunnelConfig.Index != nil {
+			index = uint32(*tunnelConfig.Index)
+		}
+		if tunnelConfig.Dir != nil {
+			dir = uint32(*tunnelConfig.Dir)
+		}
+		if tunnelConfig.HardwareID != nil {
+			hardwareID = uint32(*tunnelConfig.HardwareID)
+		}
+		writeIP(tunnelConfig.RemoteIP)
+		writeUint32(sessionID)
+		writeUint32(index)
+		writeUint32(dir)
+		writeUint32(hardwareID)
+	}
+	return fmt.Sprintf("%s-%s", portNamePrefix, hex.EncodeToString(hash.Sum(nil))[:6])
+}
+
+func ParseTrafficControlInterfaceConfig(portData *ovsconfig.OVSPortData, portConfig *interfacestore.OVSPortConfig) *interfacestore.InterfaceConfig {
+	intf := &interfacestore.InterfaceConfig{
+		Type:          interfacestore.TrafficControlInterface,
+		InterfaceName: portData.Name,
+		OVSPortConfig: portConfig}
+	if tunnelType, ok := portData.ExternalIDs[trafficControlTunnelType]; ok {
```

Comment:
If these options are must to have, you should return nil, if they are not set.

Reply:
External ID `trafficControlTunnelType` is only used when the actual type of TrafficControl target port is tunnel, like VXLAN, GENEVE, etc. If the type of TrafficControl target port is OVS internal or physical network adapter, external ID `trafficControlTunnelType` should not be set.